app/furniture/marketplace/menu_component.html.erb

```diff
@@ -1,16 +1,16 @@
-<%- marketplace.tags.group_tag.by_position.each do |tag| %>
+<%- marketplace.tags.menu_tag.by_position.each do |tag| %>
   <%- unless tag.products.empty? %>
     <h1><%= tag.label %></h1>
     <div class="grid lg:grid-cols-3 gap-3">
-      <%- tag.products.with_all_rich_text.unarchived.each do |product| %>
+      <%- tag.products.with_all_rich_text.unarchived.sort_alpha.each do |product| %>
```

Comment:
Is this type of scope chaining typical in templates? Would you put a helper on the ruby component to reduce logic in the template?

Reply:
Well, it's not too unusual, but it does make for an N+1 query. It may be worth pulling this into the query on line 1; which may be expressed in the view directly, as a helper on `MenuComponent`, or a scope on `Marketplace`.

There's also an argument to be made for not doing *any* queries from within a `Component` or view; instead performing them at the controller layer.

Reply:
What's the best way to analyze whether this lookup would produce an n+1 query? Righ now if I render out the sql in a rails console I'm not seeing multiple selects for associative rows. 

```sql
select
	"marketplace_products"."id" as t0_r0,
	"marketplace_products"."marketplace_id" as t0_r1,
	"marketplace_products"."name" as t0_r2,
	"marketplace_products"."price_cents" as t0_r3,
	"marketplace_products"."price_currency" as t0_r4,
	"marketplace_products"."created_at" as t0_r5,
	"marketplace_products"."updated_at" as t0_r6,
	"marketplace_products"."discarded_at" as t0_r7,
	"marketplace_products"."servings" as t0_r8,
	"action_text_rich_texts"."id" as t1_r0,
	"action_text_rich_texts"."name" as t1_r1,
	"action_text_rich_texts"."body" as t1_r2,
	"action_text_rich_texts"."record_type" as t1_r3,
	"action_text_rich_texts"."record_id" as t1_r4,
	"action_text_rich_texts"."created_at" as t1_r5,
	"action_text_rich_texts"."updated_at" as t1_r6
from
	"marketplace_products"
	inner join "marketplace_product_tags" on "marketplace_products"."id" = "marketplace_product_tags"."product_id"
	left outer join "action_text_rich_texts" on "action_text_rich_texts"."record_type" = 'Marketplace::Product'
		and "action_text_rich_texts"."name" = 'description'
		and "action_text_rich_texts"."record_id" = "marketplace_products"."id"
where
	"marketplace_product_tags"."tag_id" = 'b8d56ff4-8326-4a9b-8576-9e4b677bc4c4'
	and "marketplace_products"."discarded_at" is null
	/* loading for pp */
order by
	"marketplace_products"."name" asc
limit 11;
```

Reply:
An N+1 situation requires two separate queries (one that is repeated N times, plus the outer one), so you would not see an N+1 query by looking at the SQL of one query. You would need to have a look at the implementation of the `ProductComponent`, which is what you are rendering inside the loop, and see if it is going to create any additional queries.

Also, we can add https://github.com/flyerhzm/bullet to the project and help it guide us when it thinks there are N+1 queries (although note that it is not, err, bullet-proof?).

Reply:
Oh oh oh right, the risk of n+1 occurs with potential queries for each item from this collection as we loop. `bullet` gem looks useful. Sounds like similar amount of caution/skepticism we reserve for `strong_migrations`'s suggestions 😅. 

I'll take a pass in another PR adding the bullet gem and lifting this heavy lifting to a better place for Law of Demeter and n+1 code hazards.